File: 82. Remove Duplicates from Sorted List II.cs
/*A permutation perm of n + 1 integers of all the integers in the range [0, n] can be represented as a string s of length n where:

s[i] == 'I' if perm[i] < perm[i + 1], and
s[i] == 'D' if perm[i] > perm[i + 1].
Given a string s, reconstruct the permutation perm and return it. If there are multiple valid permutations perm, return any of them.

 

Example 1:

Input: s = "IDID"
Output: [0,4,1,3,2]
Example 2:

Input: s = "III"
Output: [0,1,2,3]
Example 3:

Input: s = "DDI"
Output: [3,2,0,1]
 

Constraints:

1 <= s.length <= 105
s[i] is either 'I' or 'D'.*/
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */
public class Solution {
    public ListNode DeleteDuplicates(ListNode head) 
    {
        if (head == null) return null;
        
        ListNode dummy = new ListNode();
        dummy.next = head;
        ListNode tail = dummy;
        int? prev = head.val;
        
        ListNode current = head;
        while (current != null) 
        {
            if (current.next?.val == prev)
            {
                current = current.next;
                while (current?.val == prev)
                {
                    current = current.next;
                }

                tail.next = current;
                prev = current?.val;
            }
            else 
            {
                tail = current;
                prev = current.next?.val;
                current = current.next;
            }            
        }

        return dummy.next;
    }
}

File: 82. Remove Duplicates from Sorted List II.go
/*Given the head of a sorted linked list, delete all nodes that have duplicate numbers, leaving only distinct numbers from the original list. Return the linked list sorted as well.

 

Example 1:


Input: head = [1,2,3,3,4,4,5]
Output: [1,2,5]
Example 2:


Input: head = [1,1,1,2,3]
Output: [2,3]
 

Constraints:

The number of nodes in the list is in the range [0, 300].
-100 <= Node.val <= 100
The list is guaranteed to be sorted in ascending order.
 */
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func deleteDuplicates(head *ListNode) *ListNode {
    dummy := ListNode{
        Val: -200,
        Next: head,
    }

    lastUniq := &dummy
    prev := &dummy
    node := head
    for node != nil {
        if (node.Val == prev.Val) {
            for node != nil && node.Val == prev.Val {
                node = node.Next
            }
            lastUniq.Next = node
        } else {
            lastUniq = prev
        }

        prev = node
        if (node != nil) {
            node = node.Next
        }
    }

    return dummy.Next
}
